src/contrib/mlir/compiler.cpp

```diff
@@ -238,35 +239,70 @@ MLIRCompiler::TensorInfo MLIRCompiler::get_tensor_value(descriptor::Tensor* tens
     return it->second;
 }
 
-// Lowers nGraph dialect to affine dialect.
+// Lowers nGraph dialect all the way to LLVM module.
 void MLIRCompiler::lower_ng_dialect()
 {
-    mlir::PassManager pm;
-    pm.addPass(mlir::createDialectLoweringPass(this));
-    pm.addPass(mlir::createCanonicalizerPass());
+    // Lower NG dialect to Affine
+    {
+        mlir::PassManager pm;
+        pm.addPass(mlir::createDialectLoweringPass(this));
+        pm.addPass(mlir::createCanonicalizerPass());
 
-    pm.run(m_module.get());
+        pm.run(m_module.get());
 
-    if (failed(m_module->verify()))
+        if (failed(m_module->verify()))
+        {
+            NGRAPH_CHECK(false, "Incorrect module after dialect lowering");
+        }
+
+        dump_mlir_module("Affine Dialect Dump:");
+    }
+
+    // Lower Affine to Std Dialect
     {
-        NGRAPH_CHECK(false, "Incorrect module after dialect lowering");
+        mlir::PassManager pm;
+        // Lower affine ops
+        pm.addPass(mlir::createLowerAffinePass());
+        auto rr = pm.run(m_module.get());
+        NGRAPH_CHECK(succeeded(rr), "Affine loop lowering failed");
+
+        dump_mlir_module("Standard Dialect Dump:");
```

Comment:
I'm not sure I understand why we need to change the structure of the pipeline, remove the `optimize` function and move everything into independent scopes within the same function. Is this needed to fix the recompilation issue?

Reply:
I'm actually adding optimizations to the `optimize` function (https://github.com/NervanaSystems/ngraph/pull/3290/files#diff-47415dcfbd95a002c0c4486a98db979eR271 and more locally) so if there is no strong reason for this change wrt the re-compilation issue, I would prefer we restore the original version. We can revisit this later, if needed.

Reply:
Aware of that. In your PR, you need to add `optimize_affine()` or something that will solely do the optimization (as opposed to lower + optimize)